opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -89,14 +89,28 @@ namespace Opm {
         }
         well_state_.init(wells(), cellPressures, previous_well_state_, phase_usage_);
 
+        // For each multisegment well, we need to create a mapping from perforation number from parser to
+        // the perforation number in Wells struct. In the later one, some perforations can be removed due to being
+        // completed in a inactive grid block.
+        // Ieally, it should be done in parser when parsing COMPDAT keyword. For the moment, it is done in WellsManager.
+        // This mapping information is used later when initializing the MultisegmentWells during createWellContainer.
+        std::map<std::string, std::vector<int> > perforation_mapping;
         // handling MS well related
         if (param_.use_multisegment_well_) { // if we use MultisegmentWell model
             for (const auto& well : wells_ecl_) {
-                // TODO: this is acutally not very accurate, because sometimes a deck just claims a MS well
-                // while keep the well shut. More accurately, we should check if the well exisits in the Wells
-                // structure here
-                if (well->isMultiSegment(timeStepIdx) ) { // there is one well is MS well
-                    well_state_.initWellStateMSWell(wells(), wells_ecl_, timeStepIdx, phase_usage_, previous_well_state_);
+                // there is one active well is MS well
+                if (well->isMultiSegment(timeStepIdx) && well->getStatus(timeStepIdx) != WellCommon::SHUT) {
```

Comment:
Even if we end up not using this PR, this change is an improvement we should keep.

Reply:
Sure. I will keep it. Most likely, I will open an new PR in opm-parser and update this PR.